### ISSUE TITLE
A theory regarding the missing accounts bug

### ISSUE THREAD

**Issue by juan-cortes** [LedgerHQ/ledger-live-desktop#2145]
Keep in mind this is an incomplete theory but... I got curious and started looking into the "missing accounts" issue that some of our users have been experimenting. We know that it seems to mostly happen to users who have set up a password and during an update they lose their accounts.

This points me to potentially involving the migration system where we initially had a separate file for `accounts.json` and later it was merged into a single `app.json`. 

## The migration system vs the testing system
The migration system will check a nonce variable and run all the migrations that are greater than that nonce. If you get the mock data from `userdata_v1.0.5_mock-02-encrypted-accounts`, wipe your own user data, replace it with this, and run the app, it will trigger the migration mechanism and merge all the data into `app.json`. It will all work fantastically, until it doesn't.  I was able to lose my data twice by repeating this manually for around 50 tries but we are not savages so in order to test this programmatically I ran `jest src/migrations` and it complies, it works..

## But wait!
We are debouncing the saves to the file system!, I understand this is because we have a copy in memory of the settings of the app, and we don't want a race condition or access lock affecting the `app.json` file when two or more simultaneous saves happen but when a migration is happening **we can't afford to have an active debounce in place**, or we will lose data.

<img width="822" alt="image" src="https://user-images.githubusercontent.com/4631227/60286941-f4607780-9910-11e9-9cfc-79e9cbc04fec.png">

The tests are set to only have `1ms` of debounce but the real life use to `500ms`, this means that a migration would debounce calls, when executed in the real world, if the calls are placed within half a second of eachother. If we change the `DEBOUNCE_MS` for tests to 500ms then **nothing but the window params is migrated.**

``` 
Array [
    +   "accounts.json",
        "app.json",
    -   "windowParams.json",
    +   "countervalues.json",
    +   "migrations.json",
    +   "settings.json",
    +   "user.json",
]
```

## Why does it only happen to some people
My theory is that you need a particular set of conditions for this to apply. You need enough account data for the save action to take **longer** than 500ms, so that the next call would cancel that save, so users with small accounts, small amounts of data, or super fast computers wouldn't be affected. It also means that it happens for users updating from a very old version for this to be true, so if the user has lost accounts and it's relatively new, migrations wouldn't necessarily be the cause but it could potentially be related if we are perhaps clearing the memory and saving.

Or I could be completely off, who knows 🤷‍♂ but I mean... @meriadec left a warning in the commit message for the migrations https://github.com/LedgerHQ/ledger-live-desktop/commit/4e843626178db0607dbfe331f19fd5d215aece38

## Nonce
A wrongly triggered migration would also caused the data loss, you can test this by forcing nonce to be zero, we end up writing over the whole app namespace in the storage using non existing files. The migrations should not rely on a nonce variable but rather on the files/features existing in the first place.

**Comment by gre** [comment]
This is likely plausible, I also have suspicion that the migration system tend to wrongly retrigger from time to time (maybe somehow, the nonce gets wrong? if i'm correct migration itself is stored in another file, what if this file fail to save..).

So if we confirm this with more investigation, I would be in favor of actually dropping this complex migration system.. the nonce data have to be part of the data itself. If in future we want to migrate our data, it wouldn't be too hard to just detect based on the data itself, e.g. imagine we migrate to introduce a new field: detecting presence of field can tells if migration has been done. It's like a feature detection approach and it's generally more solid than maintaining two things in sync (a nonce and the data with it).

So dropping migration system means what? We break if you come from very old version of Live I suppose? Users can reimport accounts, we will check stats on how much people would be that much outdated.

to be continued

**Comment by gre** [comment]
This is plausible, I also have suspicion that the migration system tend to wrongly retrigger from time to time (maybe somehow, the nonce gets wrong? if i'm correct migration itself is stored in another file, what if this file fail to save..).

So if we confirm this with more investigation, I would be in favor of actually dropping this complex migration system.. the nonce data have to be part of the data itself. If in future we want to migrate our data, it wouldn't be too hard to just detect based on the data itself, e.g. imagine we migrate to introduce a new field: detecting presence of field can tells if migration has been done. It's like a feature detection approach and it's generally more solid than maintaining two things in sync (a nonce and the data with it).

So dropping migration system means what? We break if you come from very old version of Live I suppose? Users can reimport accounts, we will check stats on how much people would be that much outdated.

to be continued

**Comment by meriadec** [comment]
:+1: not many things to add. We could add stress-tests, automate doing 50+ migrations using various set of debounce time, and confim what @juan-cortes manually (:joy: and courageously) tested.

So given:
- cross-platform fs stuff is weird and hard
- migrations are not really relevant anymore, we know better how to manage / structure our data and eventual changes are just additions (so: :100: for the field presence check that @gre mentionned). as a key result: we didn't even added significant commits to `migrations` folder since ~1 year (last August!).

Did we already considered solutions like [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) ? It's non-blocking, promise-based, attached directly on the browser instance, has a query system.. drawback I see is: `main` process can't access it, so we still need to store window dimensions in a file. But at least account data could be stored here.

Or we can say random data clearing it's a privacy feature. eh !

**Comment by gre** [comment]
for the window dimensions problem, we discussed the other day & i think it's technically possible to do this:
- when app boots, we don't show yet the window, instead we boot renderer in background
- the first thing renderer tells to main thread back is where to move the window
- main can then move the window to the requested position and only then do a .show()

I know it means more time to boot the app before window appear, but it's not too bad I think, we save some time for the spinner that maybe don't even need to appear (if fast computer). (Kinda like with React's suspense for the small analogy)


The first idea we have is #2018 , it can work but to an extend: there is a max local storage size. We could then split data in pieces in localStorage but that's a bit hackish. IndexedDB might have some limits too but maybe it is better. We have to assess these solution, i think it makes sense to happen in a big refactoring of our data, including removing the migrations.. except we need to migrate to indexedDB if we go there ^^

**Comment by gre** [comment]
We expect this to be fixed in context of a big rework that just got merged. we'll reopen in this does not fix in the next release #2596
